core/types/block.go

```diff
@@ -640,8 +640,43 @@ type RawBody struct {
 	Requests     Requests
 }
 
+// Represents original transaction number in blockchain (do not includes system txs in block)
+type BaseTxID uint64
+
+func CapBaseTxID(tx int) uint64 {
+	return uint64(tx) + 2
+}
+
+func (b BaseTxID) U64() uint64 { return uint64(b) }
+
+func (b BaseTxID) Bytes() []byte { return hexutility.EncodeTs(uint64(b)) }
+
+// First returns first non-system tx number in block
+// as if basetxId is first original transaction in block
+func (b BaseTxID) First() uint64 { return uint64(b + 1) }
+
+// At returns tx number at block position `ti`.
+func (b BaseTxID) At(ti int) uint64 { return b.First() + uint64(ti) }
+
+// Returns last non-system tx number in block. txns is length of block transactions (does not include system txs)
+// From the other hand, TxAmount includes 2 system txs as well as all txs in block.
+func (b BaseTxID) Last(txns int) uint64 { return b.First() + uint64(txns) }
+
+// FirstSystemTx returns first system tx number in block
+func (b BaseTxID) FirstSystemTx() BaseTxID { return b }
+
+// LastSystemTx returns last system tx number in block. result+1 will be baseID of next block a.k.a. beginning system tx number
+// eg txAmount = 3+2/*systemTx*/ = 5 therefore:
+//
+//	0 - base tx/systemBegin
```

Comment:
sustem transactions are used to write history of state changes done by consensus (not by eth-transactions) - for example "miner rewards"